### ISSUE TITLE
kbox install drupal app fails with: Container already exists error

### ISSUE THREAD

**Issue by pi-ron** [kalabox/kalabox#362]
After uninstalling / reinstalling kbox, `kbox provision`, then `kbox create drupal` runs successfully, change `kalabox.json` with settings from TL;DR guide.

This is a new app namespace, not used before.

``````
$ cd parcel-code/
$ kbox install
info: SERVICES =>  Startup services need to be started.
info: DOCKER =>  Stopping container. 3c1090d49bc441c8fd2f71816c0715fe28dd9c9d051f854d0575cc808f7b4e78
info: DOCKER =>  Container stopped. 3c1090d49bc441c8fd2f71816c0715fe28dd9c9d051f854d0575cc808f7b4e78
info: DOCKER =>  Stopping container. a499fbc57eef4076ecf34c14fe07f48d06d404a21040f8c47de32d02d61b5e16
info: DOCKER =>  Container already stopped. a499fbc57eef4076ecf34c14fe07f48d06d404a21040f8c47de32d02d61b5e16
info: DOCKER =>  Stopping container. 90dd87f5d2da5ce03f3f53fdb0ce84915c37e3da664938721be7db17173b0772
info: DOCKER =>  Container stopped. 90dd87f5d2da5ce03f3f53fdb0ce84915c37e3da664938721be7db17173b0772
info: DOCKER =>  Stopping container. e930a77abab315ec28663b5fb8c0defbbeef823690f5e91ea5fa4cccb9a86351
info: DOCKER =>  Container stopped. e930a77abab315ec28663b5fb8c0defbbeef823690f5e91ea5fa4cccb9a86351
info: DOCKER =>  Stopping container. e098829d13601619bbee79033278faa0331eb2c291cbbac67d2049eb8962b7fd
info: DOCKER =>  Container stopped. e098829d13601619bbee79033278faa0331eb2c291cbbac67d2049eb8962b7fd
info: DOCKER =>  Starting container 3c1090d49bc441c8fd2f71816c0715fe28dd9c9d051f854d0575cc808f7b4e78.
info: DOCKER =>  Container started. 3c1090d49bc441c8fd2f71816c0715fe28dd9c9d051f854d0575cc808f7b4e78
info: DOCKER =>  Starting container a499fbc57eef4076ecf34c14fe07f48d06d404a21040f8c47de32d02d61b5e16.
info: DOCKER =>  Container started. a499fbc57eef4076ecf34c14fe07f48d06d404a21040f8c47de32d02d61b5e16
info: DOCKER =>  Starting container 90dd87f5d2da5ce03f3f53fdb0ce84915c37e3da664938721be7db17173b0772.
info: DOCKER =>  Container started. 90dd87f5d2da5ce03f3f53fdb0ce84915c37e3da664938721be7db17173b0772
info: DOCKER =>  Starting container e930a77abab315ec28663b5fb8c0defbbeef823690f5e91ea5fa4cccb9a86351.
info: DOCKER =>  Container started. e930a77abab315ec28663b5fb8c0defbbeef823690f5e91ea5fa4cccb9a86351
info: DOCKER =>  Starting container e098829d13601619bbee79033278faa0331eb2c291cbbac67d2049eb8962b7fd.
info: DOCKER =>  Container started. e098829d13601619bbee79033278faa0331eb2c291cbbac67d2049eb8962b7fd
Error while installing app: "parcel-code": Container already exists: fe28747e69f0caed1422a86a6502b9f927b28c073402156e05d04d44f5df3d64
Error: Container already exists: fe28747e69f0caed1422a86a6502b9f927b28c073402156e05d04d44f5df3d64
    at /usr/local/lib/node_modules/kalabox/lib/app.js:422:17
From previous event:
    at Promise.resolve.filter.map.concurrency (/usr/local/lib/node_modules/kalabox/lib/app.js:413:8)
From previous event:
    at /usr/local/lib/node_modules/kalabox/lib/app.js:410:6
From previous event:
    at Object.exports.install (/usr/local/lib/node_modules/kalabox/lib/app.js:406:4)
    at Object.task.func (/usr/local/lib/node_modules/kalabox/plugins/kalabox-core/tasks/install.js:25:18)
    at Object.exports.run (/usr/local/lib/node_modules/kalabox/lib/core/tasks.js:773:15)
    at processTask (/usr/local/lib/node_modules/kalabox/bin/kbox.js:268:16)
    at /usr/local/lib/node_modules/kalabox/bin/kbox.js:330:19
    at process._tickCallback (node.js:419:13)
``````

**Comment by pi-ron** [comment]
No issues Pantheon app:

``````
$ kbox drush status
Identity added: /ssh/.ssh/id_rsa (/ssh/.ssh/id_rsa)
 Drupal version                  :  7.38
 Site URI                        :  localhost
 Database driver                 :  mysql
 Database username               :  pantheon
 Database name                   :  pantheon
 Database                        :  Connected
 Drupal bootstrap                :  Successful
 Drupal user                     :  Anonymous
 Default theme                   :  westpac_rescue
 Administration theme            :  ember_sub
 PHP executable                  :  /root/.phpbrew/php/php-5.3.29/bin/php
 PHP configuration               :  /root/.phpbrew/php/php-5.3.29/etc/php.ini
 PHP OS                          :  Linux
 Drush version                   :  6.6.0
 Drush configuration             :
 Drush alias files               :  /root/.drush/aliases.drushrc.php
                                    /root/.drush/pantheon.aliases.drushrc.php
 Drupal root                     :  /code
 Site path                       :  sites/default
 File directory path             :  sites/default/files
 Private file directory path     :  sites/default/files/private
 Temporary file directory path   :  /tmp
``````

**Comment by pirog** [comment]
do the two namespaces have anything in common?

**Comment by pi-ron** [comment]
They use the same codebase / profile

**Comment by pirog** [comment]
i mean do the app names have anything in common aka "bobapp1" and "bobapp2" share the name "bobapp"

**Comment by pi-ron** [comment]
Yup:
parcel-upstream
parcel
parcel-code

**Comment by pi-ron** [comment]
I've now used a Pantheon app, created on Pantheon first then pulled down as `parcel-dev`

**Comment by pirog** [comment]
try creating an app that doesnt start with the word "parcel" aka something
like "thing" and see if you get the same issue.

On Thu, Aug 6, 2015 at 9:39 AM, pi-ron <notifications@github.com> wrote:

> I've now used a Pantheon app, created on Pantheon first then pulled down
> as parcel-dev
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kalabox/kalabox/issues/362#issuecomment-128205896>.
>



-- 

Mike

-- 
Mike Pirog
Co-Founder, Products
http://kalamuna.com

Stay connected with Kalamuna on: Twitter
<http://t.sidekickopen01.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJW7t5XZs1pwYhxW8qlTmx64kckdW8pTc_456dT6hf3BRzQ602?t=http%3A%2F%2Ftwitter.com%2Fkalamuna&si=5428072445640704&pi=f16b515d-1f50-46c7-aebf-11a0c79916b9>
, Linkedin
<http://t.sidekickopen01.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJW7t5XZs1pwYhxW8qlTmx64kckdW8pTc_456dT6hf3BRzQ602?t=http%3A%2F%2Fwww.linkedin.com%2Fcompany%2F2584606&si=5428072445640704&pi=f16b515d-1f50-46c7-aebf-11a0c79916b9>,
and Facebook
<http://t.sidekickopen01.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJW7t5XZs1pwYhxW8qlTmx64kckdW8pTc_456dT6hf3BRzQ602?t=http%3A%2F%2Fwww.facebook.com%2Fkalamunatech&si=5428072445640704&pi=f16b515d-1f50-46c7-aebf-11a0c79916b9>

**Comment by pi-ron** [comment]
No luck there unfortunately:

``````
Error while installing app: "nugget": Container already exists: 840dbd49f092ff8d1b326a929b21eaab8626e4727bf9b590a12b4eb6c46e39cf
Error: Container already exists: 840dbd49f092ff8d1b326a929b21eaab8626e4727bf9b590a12b4eb6c46e39cf
    at /usr/local/lib/node_modules/kalabox/lib/app.js:422:17
From previous event:
    at Promise.resolve.filter.map.concurrency (/usr/local/lib/node_modules/kalabox/lib/app.js:413:8)
From previous event:
    at /usr/local/lib/node_modules/kalabox/lib/app.js:410:6
From previous event:
    at Object.exports.install (/usr/local/lib/node_modules/kalabox/lib/app.js:406:4)
    at Object.task.func (/usr/local/lib/node_modules/kalabox/plugins/kalabox-core/tasks/install.js:25:18)
    at Object.exports.run (/usr/local/lib/node_modules/kalabox/lib/core/tasks.js:773:15)
    at processTask (/usr/local/lib/node_modules/kalabox/bin/kbox.js:268:16)
    at /usr/local/lib/node_modules/kalabox/bin/kbox.js:330:19
    at process._tickCallback (node.js:419:13)
``````

**Comment by pirog** [comment]
"nugget" lol. 

what directory are you in when you are creating the other app? are you in another app directory?

**Comment by pi-ron** [comment]
The golden nugget.

No in ~/Sites

**Comment by pirog** [comment]
@clickbox, does this look about what you were getting to? could you write out the steps you took to  produce this?

**Comment by clickbox** [comment]
@pirog Yes this is what was happening to me before I uninstalled virtualbox with the uninstall script. To my knowledge I produced this by deleting a app's directory folder before shutting down the containers.

**Comment by pirog** [comment]
@pi-ron is the folder name that holds your app the same as the name of the app itself?

**Comment by pi-ron** [comment]
@pirog yes folder name is always the same as app name

**Comment by pirog** [comment]
@pi-ron yeah i tried this out as well and having different names doesnt seem to matter.

**Comment by pi-ron** [comment]
Is everyone having this issue or just me?

**Comment by reynoldsalec** [comment]
I have experienced something like this in the far past @pi-ron when I deleted an app before shutting it down (perhaps similar to what @clickbox experienced). I then started getting the error on every app.

I think I ultimately resolved things by running the uninstall script, although before that I tried deleting the old containers manually. Could be worthwhile to see if there are any legacy containers stalking around your install.

**Comment by clickbox** [comment]
sounds exactly the same @reynoldsalec. the container was a ghost to me. Shouldn't that hash of the "existing" container be able to get removed with docker rm ? I'm almost wondering if I should have tried to remove it outside of boot2docker =vm"Kalabox2" ssh

**Comment by pirog** [comment]
@pi-ron @clickbox back on the scene and i think this will be slotted for next sprint.

**Comment by pirog** [comment]
I can't seem to replicate this on the latest version so going to backlog until i can figure that out.

**Comment by pirog** [comment]
Ok, so as far as i can tell this happens when you delete an app directory without doing a `kbox destroy` first. I'm honestly not sure how we can guard against this though :)

**Comment by rgeleta** [comment]
Have similar problem, and do know I deleted an app directory without doing a `kbox destroy` first.

Understand you cannot guard against this, but it would be helpful to have instructions to manually clean up the self created mess.

**Comment by pirog** [comment]
Better documentation would be good. 

That said i think there are some somewhat related issues floating around. Namely this  and #456. 

I think a TOTAL RESOLUTION for both could be:

1. Document app destroy better
2. Detect if a container is an orphaned container and then allow it to be replaced (that would resolve this issue)
3. Have `kbox destroy` remove the ENTIRE app directory
4. Make sure we are cleaning up the appRegistry better.

**Comment by rgeleta** [comment]
speaking for apparently more than myself
in our defense, README.md has
`cd /dir/i/want/my/app/to/live`

it is a reasonable (albeit wrong) assumption that the app would live in `cd /dir/i/want/my/app/to/live` and if we deleted that we would delete the app

what we do (did?) not know is that there is such a thing as an appRegistry, or that supporting software may not be designed or have the capability to be told and honor a request to put their stuff somewhere other than where they do, and kalabox does the best it can with what it uses.

just sayin

**Comment by pirog** [comment]
Hey @rgeleta, to clarify my proposed total resolution is a list of things KALABOX can and should do in the future to alleviate your concerns. This would include being able to "delete" an app just by deleting the directory in which it lives. 

You don't need to worry about things like the appRegistry. Those are "system" files that we manage on our end. However, it lives at `~/.kalabox/appRegistry.json`.

**Comment by pirog** [comment]
Going to add some notes in here:

**Scenario 1:**

```
kbox create pantheon
cd myApp
kbox start
cd ..
rm -rf myApp
```

Kalabox recognizes the app deletion from the file system and removes it from the appRegistry. This means we just need to handle the orphaned containers. We can do two things here

1. When the appRegistry change is noted we autodestroy (this would guarantee that the app's containers are stopped, syncthing changes are made and we dont have orphaned containers)
2. Keep containers orphaned until user tries to create another app with the same. (Advantage here is less obtuse ux)

**Comment by pirog** [comment]
So after discussing with @bcauldwell we are going to go with 1 from above. The details are

1. Create a maintenance promise inside of app.js to run in parallel during app.install
2. Method will do the following
  a. Compare appRegistry vs. app.list and note difference
  b.If different remove needed entries from appRegistry with new disco.removeEntry func
  c. Do a kbox destroy to make sure any orphaned containers and syncthing folders are put into the proper state.

**Comment by pirog** [comment]
So this is actually going to require a much deeper change :(

1. We need to move the `cids` out of the `:appRoot:` and into the `:sysConfRoot:` so we don't have to worry about the user removing them. This would also require we do some sort of update process for users who have their `cids` still in the `:appRoot:`.
2. Once we make this change we can identify apps with "orphaned" containers easily because they will be apps that don't exist in `kbox.app.list()` but have `cids` in the `:sysConfRoot`. We also know that apps in the `appRegistry` that are not in `kbox.app.list()` and that do not have `cids` in `:sysConfRoot:` can be safely removed from `appRegistry.json`
3. It becomes trivial to generate a stock appConfig for orphaned containers which means we can remove these apps with the normal assortment of app commands.

**Comment by pirog** [comment]
Adding a note here:

1. If you run `kbox update` and you have an "orphaned" app the whole thing fails because syncthing cant find the codeRoot of a deleted app.

**Comment by pirog** [comment]
Update on progress here

1. ~~We need to move the `cids` out of the `:appRoot:` and into the `:sysConfRoot:` so we don't have to worry about the user removing them. This would also require we do some sort of update process for users who have their `cids` still in the `:appRoot:`.~~
2. Once we make this change we can identify apps with "orphaned" containers easily because they will be apps that don't exist in `kbox.app.list()` but have `cids` in the `:sysConfRoot`. We also know that apps in the `appRegistry` that are not in `kbox.app.list()` and that do not have `cids` in `:sysConfRoot:` can be safely removed from `appRegistry.json`
3. It becomes trivial to generate a stock appConfig for orphaned containers which means we can remove these apps with the normal assortment of app commands.

I think https://github.com/kalabox/kalabox/issues/362#issuecomment-145719690 will be resolved by completing the above

**Comment by pirog** [comment]
Next Steps

1. Have kbox destroy remove the ENTIRE app directory
2. Have kbox destroy tell the user WTF its gonna do, confirm required
3. Add `app.cleanup` to `kbox update`, `kbox start`, `kbox stop`, `kbox install`, others?

**Comment by pirog** [comment]
This issue should be resolved in the new 0.10.3 release!
